Image block: Add aspect ratio support to lightbox

artemiomorales opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This PR makes it so the lightbox zoom animation functions as expected when configuring a custom aspect ratio on an image.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
Addresses #52192
We want the lightbox to be compatible with all of the image block's features.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
It calculates the aspect ratio of the resized image and uses it to calculate the target zoom dimensions.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->
1. Ensure interactivity API experiments are enabled
2. Add an image to a post
3. Customize the image's `aspect ratio` in block settings
4. Activate the lightbox under Advanced > Behaviors > Lightbox. Make sure the `zoom` animation is selected.
5. Publish, view the post, and click on the image — it should zoom in as expected.

*If possible, please also test on a mobile device.*

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->
N/A

## Screenshots or screencast <!-- if applicable -->

### Before

https://github.com/WordPress/gutenberg/assets/5360536/08c395b4-afbb-4a16-a518-034d8935ea8c

### After

https://github.com/WordPress/gutenberg/assets/5360536/c66020b0-8e75-4187-aefe-8007c2ac5dcb

SantosGuillamot committed:
Initial aspect-ration support implementation

SantosGuillamot committed:
Move `inherit` property to `wp-style` directive

SantosGuillamot committed:
Add `object-fit` cover

SantosGuillamot committed:
Fix big images aspect ratio

artemiomorales committed:
Fix animation on mobile; add comments

artemiomorales commented:
I can't think of any automated tests to go with this one, please just let me know if any make sense!

artemiomorales committed:
Remove obsolete variable declaration

artemiomorales committed:
Fix PHP spacing

SantosGuillamot commented:
As we worked together on this one, I'd prefer if @michalczaplinski or @c4rl0sbr4v0 could take a look at the code for the review 🙂

cbravobernal commented:
The image flickers on Safari browser 😿 

https://github.com/WordPress/gutenberg/assets/37012961/fcc60073-9bfb-4596-a324-7d725fbc3a4a

artemiomorales committed:
Optimize animation performance; add comments; fix fade and thumbnail bugs

artemiomorales commented:
> The image flickers on Safari browser 😿

@c4rl0sbr4v0 Thanks for catching this! The latest push fixes this jitteriness in Safari, as well as cleans up the animation code and addresses the use of a regular `thumbnail` resolution, which I realized wasn't working.

However, while working on all of that, I noticed a couple of other cases we still haven't accounted for:

- If a user selects the `thumbnail` resolution, then selects a custom aspect ratio, the zoom animation does not work properly
- We haven't provided support for when a user selects `contain` for the aspect ratio scale

In addition, something in my code is causing the PHP linter to fire an error.

Will continue hammering at these issues.

SantosGuillamot commented:
I guess the different devices interpret the CSS animations differently, right? It already happened that, using `translate(-50%, -50%)`, wasn't working properly in Android devices. Maybe we can check how other solutions are creating this zooming animation.

michalczaplinski commented:
> If a user selects the thumbnail resolution, then selects a custom aspect ratio, the zoom animation does not work properly

Is this the issue that you mention @artemiomorales ? Then, I can reproduce it too:

https://github.com/WordPress/gutenberg/assets/5417266/3abbc816-eeaa-4854-8a62-eaff3d1a747b

michalczaplinski reviewed (commented):
I think I'm not too familiar with this part of the code to be able to review the code changes properly. A short video to explain what's going on could be useful.

Otherwise, I've tested the PR and it works as expected (sans the issue with thumbnails that I mention above)! 👍

artemiomorales commented:
> Is this the issue that you mention @artemiomorales ? Then, I can reproduce it too:

@michalczaplinski Yup, that's the one!

> I guess the different devices interpret the CSS animations differently, right? It already happened that, using translate(-50%, -50%), wasn't working properly in Android devices. Maybe we can check how other solutions are creating this zooming animation.

@SantosGuillamot So the issue with the thumbnails is that, when one selects `thumbnail` as an option, WordPress will serve a 1:1 aspect ratio source image, whereas the full-sized source image in the lightbox is typically *not* be 1:1. Moreover, if one specifies a custom aspect ratio, the image in the content will be resized accordingly within the container, but still using the 1:1 source image.

So as far as I see, to get this to work, we need to somehow calculate CSS values for the the full-sized image that would create the initial starting conditions of the thumbnail image while having completely different dimensions. There's probably a way to do it — am still thinking through it. 

There may also be a way to handle the UX differently, but this seems to me to the approach that users would expect. Am happy to hear any ideas or suggestions! Though this case seems pretty specific for our implementation and constraints, I'll see about investing some time into looking at additional zoom implementations too.

artemiomorales committed:
Handle images in content with aspect ratios that differ from source images

Note: This commit does not yet contain support for images with the
`contain` style enabled, or for the images with 16x9 aspect ratio.

For the moment, This code works by checking the aspect ratio of
the image's natural dimensions and comparing it to the target dimensions.

If these aspect ratios differ, then we calculate the space the lightbox
image would use as a cover image and apply that as a `scale` CSS attribute.

It also currently sets the image as a square using that resized dimension,
which seems to work in many but not all cases, so will continue to explore
the best way to approach.

Animation for translation and scale have been separated in this commit
to allow for flexibility in handling various scenarios.

artemiomorales commented:
EDIT: Updated to specify that that 16x9 aspect ratio problem is in the special case where a image has set to thumbnail resolution with a custom aspect ratio of 16x9.

The latest push seems to handle most cases, though it doesn't play nicely when a thumbnail has been defined with a 16x9 aspect ratio; I also need to add support for the `contain` image block setting.

For the moment, it checks the aspect ratio of the image's natural dimensions (for thumbnails, this is 150x150), and compares it to the aspect ratio of the target dimensions (this is the full-sized image).

If these aspect ratios differ, then we calculate the space the lightbox image would use as a cover image and apply that as a `scale` CSS attribute.

The logic isn't quite there yet and needs to be cleaned up, but would appreciate any suggestions and thoughts!

cbravobernal commented on packages/block-library/src/image/style.scss:
```diff
@@ -196,24 +197,32 @@
+
+		@media screen and (min-width: 1920px) {
+			padding: 40px 80px;
+		}
```
Margins are paddings are usually defined in styles, can we move this to CSS vars calculations?

![Screenshot 2023-07-25 at 11 15 42](https://github.com/WordPress/gutenberg/assets/37012961/03b4962b-fcfb-4f39-bd76-e58e8e444a96)

SantosGuillamot commented:
I've been testing it, and the 16x9 aspect ratio seems to work for me. What issue are you having?

Apart from that, I guess that for the contain setting we could create something like this?

```js
let originalWidth = offsetWidth;
let originalHeight = offsetHeight;
let originalLeftPosition = screenPosX;
let originalTopPosition = screenPosY;

// If the image has "object-fit: contain", calculate the dimensions without the extra space.
if (event.target.nextElementSibling.style.objectFit === "contain") {
  // If the image has blank space in the height, keep the width and calculate the height based on the natural ratio.
  if (naturalRatio > offsetRatio) {
    originalHeight = offsetWidth / naturalRatio;
    const extraHeight = offsetHeight - originalHeight;
    originalTopPosition = originalTopPosition + extraHeight / 2;
  } else {
    // If the image has blank space in the width, keep the height and calculate the width based on the natural ratio.
    originalWidth = offsetHeight * naturalRatio;
    const extraWidth = offsetWidth - originalWidth;
    originalLeftPosition = originalLeftPosition + extraWidth / 2;
  }
}

```

On another note, I would like to review the whole zooming implementation deeper. I feel that for each edge case we found (like the Android issue, Safari issue, the aspect ratio, the contain functionality, etc) we are adding more JavaScript to handle each case, and the logic is becoming quite complex. I want to fully understand why those issues are happening, the reasoning behind the JS logic, and ensure it cannot be simplified and moved part of it to CSS.

SantosGuillamot committed:
Simplify zooming lightbox

SantosGuillamot committed:
Add support for thumbnails

SantosGuillamot committed:
Clean code and change variable names

artemiomorales commented:
I merged in changes by @SantosGuillamot refactoring the implementation. It's looking much better! I tested on a couple of mobile devices and it largely functions as expected. We do still have a few scenarios to cover:

1.) This portrait image with `thumbnail` resolution, aspect ratio of 9:16, and `cover` doesn’t animate properly, which maybe points at some faultiness in the logic somewhere that could surface with other images.

https://github.com/WordPress/gutenberg/assets/5360536/91542bf0-edcf-48ec-8790-72f69880abae

2.) Thumbnails with `contain` don’t function well yet. Sometimes the lightbox image is the wrong aspect ratio, at other times there’s an erroneous overflow.

https://github.com/WordPress/gutenberg/assets/5360536/1dcedd24-7315-40a4-b605-0582294a7b13

3.) In general, we need to fix the inconsistent behavior with `contain`. As a user, I’d expect the lightbox image to display at full size when opened regardless of having `cover` or `contain` selected.

4.) We need to add the padding back in.

5.) The responsive image beneath the large image often shows as a broken image link.

SantosGuillamot commented:
Thanks for testing and raising the issues!

I didn't consider that an image could be cropped both in width and height. I just made a commit to fix that use case (I believe).


https://github.com/WordPress/gutenberg/assets/34552881/c6be93cc-0aa8-4deb-9b69-cf4acb123208

It'd be great if you can test it as it depends on the image resolution.

I'll take a look at the issue with contain.

SantosGuillamot committed:
Fix thumbnails with cropped width and height

SantosGuillamot committed:
Add support for contain setting

SantosGuillamot commented:
I added support for the contain setting in the latest commit. I've tested it in multiple scenarios, and it seems to work. But it'd be great if you could test it with different images.



https://github.com/WordPress/gutenberg/assets/34552881/85c766c9-6623-446c-92fd-4d721402e3f0

SantosGuillamot committed:
Add 5% padding to the lightbox container

SantosGuillamot committed:
Move CSS properties to style tag

SantosGuillamot committed:
Add wp prefix to CSS global variables

SantosGuillamot committed:
Remove `!important` from CSS

SantosGuillamot commented:
I rebased trunk and I added a couple of commits to clean the code and remove the `!important` from CSS.

On another note, I just realized that the fade in animation is not working the same way for the different uses cases. I believe we can reuse the zooming logic. I'll try that.

SantosGuillamot committed:
Reuse zooming logic for the fade animation

SantosGuillamot commented:
I added a commit to reuse the zooming logic for the fade animation. It seems to solve all the problems the fade animation had with the different aspect ratio, contain, and thumbnails. However, I'm not 100% sure why there was a logic specific to the zoom, so I may be missing something.

SantosGuillamot committed:
Remove fade conditional

cbravobernal committed:
Fix php standards

SantosGuillamot committed:
Remove unnecessary inheritSize selector

cbravobernal committed:
Update e2e test to fit new style tag

artemiomorales committed:
Fix width of image container so button isn't wider than image

artemiomorales committed:
Add variable horizontal padding and fixed vertical padding

artemiomorales committed:
Fix fade out animation; modify animation for image slightly

artemiomorales commented:
@SantosGuillamot Great, thanks! I fixed a few final small issues:

1.) When using `contain`, the button to open the lightbox was rendering as wider than the image in the content — that's been fixed so the button is only as wide as you'd expect.

2.) I revised the padding in accordance with the [design](https://github.com/WordPress/gutenberg/pull/51357#issuecomment-1588865012), namely doing a horizontal padding of 0px on mobile, 40px on tablet, and 80px on desktop, and a fixed vertical padding of 40px.

3.) While the fade was generally working as expected, the _fade out_ animation had disappeared, so I added that back in, as well as a little polish with fading the image slightly slower or faster than the scrim.

I ran across a few additional issues, none of which I believe are worth holding up this PR, but are worth mentioning:

- When doing using the fade, the scroll position of the window sometimes changes unexpectedly. It's particularly egregious in this example, but less so with most images that I see:

https://github.com/WordPress/gutenberg/assets/5360536/bcb711be-dafe-4e63-ad29-436eebf88208

- When using `contain` and the image leaves visible space on the top and bottom, the button to open the lightbox is actionable in that visible space, which is unexpected. This is slightly different than the case I fixed above, and would probably require some additional calculations to get right, as the image's height and width appear to be the size of the container.

https://github.com/WordPress/gutenberg/assets/5360536/6598ac38-cf3f-43fe-aeca-174f69ddf149

- In desktop Chrome and Firefox, on narrow widths, sometimes the zoom out animation doesn't behave as expected. Interestingly, this is bug is present on desktop but not on actual mobile devices which actually use that width, so I'm not sure this is worth fixing (or at least, I think it should be lower priority)

https://github.com/WordPress/gutenberg/assets/5360536/e627c6b5-f87a-46be-9deb-30378689cae7

That being said, from what I can see this PR is good to go and we can file issues for the points above — am happy to hear what others think. I did test on Android and iOS, and things are looking pretty good good. For large portrait images on Android using Edge, the image is sometimes cut off on the bottom, so that may be worth exploring.

artemiomorales commented:
> However, I'm not 100% sure why there was a logic specific to the zoom, so I may be missing something.
 
I believe the reason has been resolved by the refactor, so we're good to go with the current implementation 😄

artemiomorales commented:
> I added support for the contain setting in the latest commit. I've tested it in multiple scenarios, and it seems to work. But it'd be great if you could test it with different images.

@SantosGuillamot I actually haven't gotten a chance to test with different images — can take a look next week.

SantosGuillamot commented:
Thanks for taking a look at it @artemiomorales ! I can take a look at the issues you mentioned deeper. I tried to reproduce the one with the scroll and the zooming out but I wasn't able to reproduce it.

Apart from that, I was discussing the other day with @c4rl0sbr4v0 and I am not sure if we should download the lightbox images on hover. In large pages with many images we could be downloading much more images than necessary. Additionally, I believe we should add some logic to identify if that image has been downloaded before. I made a video explaining the problem I see and how I believe it could be fixed:

https://www.loom.com/share/0228089a37b6441babace0de33c4abe9

And this is the branch with the code: [Link](https://github.com/WordPress/gutenberg/tree/update/lightbox-download-image-on-click).

As mentioned in the video, I believe we could simplify the logic a bit as well.

SantosGuillamot committed:
Prevent scroll after focusing last image

SantosGuillamot commented:
Regarding the issues you mentioned:

1 - Scroll position change after fade out animation.

I've been able to reproduce this one. With vertical images, this bug is especially annoying. If I am not mistaken, this is caused by the `focus` function we are calling to set the focus on the previous image. Fortunately, I believe it can be solved using `preventScroll: true`. I made this change in this commit and it seems to work:


https://github.com/WordPress/gutenberg/assets/34552881/6909bfd7-fd40-44d7-8734-4e505226428e

2- When using `contain`, the button to open the lightbox is actionable in that visible space:

Technically, the blank space is part of the image and I guess that's why it is part of the button as well. I believe we could make the button fit the real image with a `wp-effect`. I can try it.

3 - In desktop Chrome and Firefox, on narrow widths, sometimes the zoom out animation doesn't behave as expected.

I haven't been able to reproduce this one. Could you test if removing the `overflow: hidden`, as Carlos did in [this PR](https://github.com/WordPress/gutenberg/pull/53026/files), solves the issue for you?

SantosGuillamot committed:
Add support for contain in lightbox image button

SantosGuillamot commented:
Regarding the second issue, I added [this commit](https://github.com/WordPress/gutenberg/pull/52765/commits/a7a2875641cd999d3cc1570c7a6b30e360f52cc7) to make the button the same size as the real image. It'd be great if you can double-check that it works as expected.

artemiomorales commented:
> Apart from that, I was discussing the other day with @c4rl0sbr4v0 and I am not sure if we should download the lightbox images on hover. In large pages with many images we could be downloading much more images than necessary. Additionally, I believe we should add some logic to identify if that image has been downloaded before. I made a video explaining the problem I see and how I believe it could be fixed:
>
> https://www.loom.com/share/0228089a37b6441babace0de33c4abe9

@SantosGuillamot Thanks for taking a look at this! We can remove the loading of the lightbox on hover; its improvement to the UX is minor, so no big issues there.

We actually don't need to add logic to prevent the image from being loaded multiple times — the browser cache will automatically prevent that from happening.

As for cleaning up the code, I gave it a go and initially thought that using a combination of the `lazy` and `hidden` attributes didn't work, but now think it might — I'll take another look tomorrow to double check.

artemiomorales commented:
> 1 - Scroll position change after fade out animation.
>
> I've been able to reproduce this one. With vertical images, this bug is especially annoying. If I am not mistaken, this is caused by the focus function we are calling to set the focus on the previous image. Fortunately, I believe it can be solved using preventScroll: true. I made this change in this commit and it seems to work:

@SantosGuillamot  Great, I believe this has been fixed.

> 2- When using contain, the button to open the lightbox is actionable in that visible space:
> 
> Technically, the blank space is part of the image and I guess that's why it is part of the button as well. I believe we could make the button fit the real image with a wp-effect. I can try it.

This issue is still present for images that fall outside of the viewport, because the `naturalHeight` and `naturalWidth` for those elements are not yet defined. I believe we'll need to add a `scroll` event listener to handle this use case. In doing so, I've been running into issues with the Interactivity API directives. Will take a closer look tomorrow.

> 3 - In desktop Chrome and Firefox, on narrow widths, sometimes the zoom out animation doesn't behave as expected.
>
> I haven't been able to reproduce this one. Could you test if removing the overflow: hidden, as Carlos did in [this PR](https://github.com/WordPress/gutenberg/pull/53026/files), solves the issue for you?

Removing `overflow: hidden` does indeed solve the issue, but I'd like to wait for feedback from the accessibility reviewers before moving forward with that as a fix, as I believe there may be issues with screenreaders, overlays, and overflow that we may need to account for.

SantosGuillamot commented:
> This issue is still present for images that fall outside of the viewport

You're right. Maybe we can get that value from the server directly. It seems that using something like [wp_get_attachment_image_src](https://developer.wordpress.org/reference/functions/wp_get_attachment_image_src/) is worth exploring. I would try to avoid adding a `scroll` event listener if that's triggered each time the user scrolls.

artemiomorales committed:
Add logic to set button styles only when needed and on image load

artemiomorales committed:
Simplify lazy loading logic

artemiomorales commented:
> You're right. Maybe we can get that value from the server directly. It seems that using something like [wp_get_attachment_image_src](https://developer.wordpress.org/reference/functions/wp_get_attachment_image_src/) is worth exploring. I would try to avoid adding a scroll event listener if that's triggered each time the user scrolls.

@SantosGuillamot Thanks for the suggestion! I ended up adding a [callback on image load](https://github.com/WordPress/gutenberg/pull/52765/commits/fa63a2f490e58fed693d2804d73545e1ccbf066b) to prevent unnecessary processing on the server. With this, all of the `contain` use cases have been handled as far as I can tell.

I also retested some of your lazy loading logic and it turns out we could in fact simplify that a bit. I ended up removing the selectors for the responsive and enlarged images, and the code is cleaner now.

With that, barring any additional issues, I believe this PR is ready to go. I rebased on trunk, so we'll see if the tests pass. I'll also do another round of mobile testing tomorrow to make sure none of these last changes broke anything.

Am happy to hear any additional ideas or insights on this!

SantosGuillamot commented:
I've been testing it, and I realized that, with the current implementation, the Full Width and Wide Width options (at least) are broken. Taking a quick look it seems the problem is caused because we are adding a extra div with some CSS. The property breaking it is `width: max-content`: [link](https://github.com/WordPress/gutenberg/blob/ec5f76a035b9f1ff8990758a0d3654a10e352947/packages/block-library/src/image/style.scss#L158).

I am a bit afraid that, in the same way it affects the Full Width and Wide Width implementation, it could affect custom CSS the users are adding. The image has been there for a while, so there might be multiple custom implementations.

Ideally, I believe we should try to remove the extra `div` and keep the same HTML but with the extra button with `position: absolute`. If I am not mistaken, that shouldn't break current implementation. Now that we are calculating the width and height with JavaScript, maybe it is feasible. I quickly tested it by adding the following CSS to the button and `position: relative` to the parent figure:

```CSS
top: 50%;
left: 50%;
transform: translate(-50%, -50%);
```

However, we might need to calculate the positioning as well because each case might be different.

Additionally, I believe we should aim not to add anything to the figure at all because it could cause unexpected results in users relying on the current implementation.

artemiomorales committed:
Remove extra div from image markup

As part of this commit, I added handling for button styles
in all images with the lightbox enabled, as we can't rely
on the CSS as was written to set the right dimensions
without breaking the layout on mobile.

artemiomorales committed:
Add 1 pixel to container dimensions to fix style bug on iOS

artemiomorales commented:
> Ideally, I believe we should try to remove the extra div and keep the same HTML but with the extra button with position: absolute. If I am not mistaken, that shouldn't break current implementation. Now that we are calculating the width and height with JavaScript, maybe it is feasible. 

@SantosGuillamot  Thanks for catching this. I've removed the extra div as you've suggested along with a couple of other modifications and everything seems to be working as far as I can tell. Next step will update the tests.

When using `full width`, there is an issue with the zoom out animation wherein the image doesn't center properly, which  can be resolved by removing `overflow: hidden` on the root HTML element as @c4rl0sbr4v0 is exploring in #53026, so will continue looking at that issue there.

artemiomorales committed:
Add logic to center button when using 'contain'

artemiomorales commented:
The [latest commit](https://github.com/WordPress/gutenberg/pull/52765/commits/282e22a79d67a960167669dd54f07c2d41beda30) adds logic to recalculate the button styles on window resize, as per a suggestion from @luisherranz.

It includes new logic for executing logic after hydration with the Interactivity API, based on this [conversation of an `afterLoad` function](https://github.com/WordPress/gutenberg/discussions/52514#discussioncomment-6476546), which I use to register a `resize` callback.

I also added a `debounce()` function to make sure we aren't executing the `resize` callback too often.

In addition to asking for general feedback, I have a few questions for discussion:
1. Should the additional logic for the Interactivity API be a part of this PR or a separate one?
2. Should we include a way to add cleanup code to that logic?
3. Do we have an existing `debounce()` function in the codebase I should use?
4. How do you think we should subscribe the `setButtonStyles` method to the state? Right now I just add a reference to `state.windowWidth` and `state.windowHeight` in a conditional, which does the trick.

luisherranz reviewed (changes requested):
We are going to move the introduction of `afterLoad` to its own PR, so please, wait until that other PR is merged to merge this one. We'll add a link here once it's opened.

SantosGuillamot commented:
I tested it and it seems to be working great! 👏 

I wonder if we could do something similar for the [`setStyles` function](https://github.com/WordPress/gutenberg/blob/282e22a79d67a960167669dd54f07c2d41beda30/packages/block-library/src/image/view-interactivity.js#L284-L447). The styles of the lightbox also depend on the `window.width` and `window.height`. Maybe we could add an effect like `setButtonStyles` but for this?

artemiomorales commented:
> I wonder if we could do something similar for the [setStyles function](https://github.com/WordPress/gutenberg/blob/282e22a79d67a960167669dd54f07c2d41beda30/packages/block-library/src/image/view-interactivity.js#L284-L447). The styles of the lightbox also depend on the window.width and window.height. Maybe we could add an effect like setButtonStyles but for this?

@SantosGuillamot  Hmm, that seems like a good idea on the surface, but there's no need for `setStyles` to subscribe to the `resize` event since it gets fired explicitly when the lightbox button is clicked.

Maybe it would be cleaner to pass the state as an argument to read the window dimensions from it, but there's no technical reason to do this and I'm not sure it would improve the readability of the code.

Do you think there'd be any benefit to doing this? Am happy to discuss further 😄

SantosGuillamot commented:
> Do you think there'd be any benefit to doing this? Am happy to discuss further 😄

I suggested it because if you open the lightbox, and resize the window while the lightbox is opened, it doesn't work as expected.

https://github.com/WordPress/gutenberg/assets/34552881/d14f5763-b643-4401-87c9-fb28a27ead35

It's not a huge problem because it only happens when the lightbox is opened and, if you close it and open it again it works. If it's solved just by moving the logic to an effect that listens to the global state, maybe it was worth trying. But you have much more context than me 🙂

artemiomorales commented:
> I suggested it because if you open the lightbox, and resize the window while the lightbox is opened, it doesn't work as expected.

Good point! Yes then I think adding support for resizing the screen while the lightbox is active would be a nice addition. I'll add that to the list of tasks.

luisherranz commented:
For reference, this is the PR for the `afterLoad`:

- https://github.com/WordPress/gutenberg/pull/53363

artemiomorales committed:
(After merge) Add logic to set button styles on window resize

DAreRodz committed:
Add docs for `afterLoad`

DAreRodz committed:
Move store options to the end

artemiomorales committed:
Replace loading=lazy with manual setting of src attribute

The combination of 'loading=lazy' and the 'hidden' attributes
was breaking progressive image loading in Safari and Chrome; instead
we are now taking what seems to be a more reliable approach
to accomplish lazy loading of the enlarged image, namely
setting the src manually when the lightbox is opened.

*This was an approach we had implemented before, but we began using
'loading=lazy' during a code cleanup.

artemiomorales commented:
While testing, I realized that using `loading="lazy"` in combination with a `hidden` attribute will prevent progressive loading from happening in Safari and Chrome.

On the other hand, *not* using the lazy attribute and instead populating the src attribute manually when the lightbox opens behaves as expected.

### Without `lazy` attribute

https://github.com/WordPress/gutenberg/assets/5360536/197f9bf5-90d3-4e31-af3a-e7d352e1ce5f

### With `lazy` attribute

https://github.com/WordPress/gutenberg/assets/5360536/83ee80fa-01c0-4bbf-8fa1-6ec3eb999046

With that in mind, I'll revise the implementation to remove the `lazy` attribute and set the `src` manually, which was how we were handling the load of the large image in an earlier version of this PR. Am happy to hear additional thoughts!

artemiomorales committed:
Add link to comment

artemiomorales committed:
Remove obsolete code

artemiomorales committed:
Update tests

artemiomorales commented:
> It's not a huge problem because it only happens when the lightbox is opened and, if you close it and open it again it works. If it's solved just by moving the logic to an effect that listens to the global state, maybe it was worth trying. But you have much more context than me 🙂

@SantosGuillamot How about we save this for a follow up PR? I started looking at this and, while I think it's quite feasible, it would require us to move some things around and would be much easier to implement and review separately.

> For reference, this is the PR for the afterLoad:
>
> https://github.com/WordPress/gutenberg/pull/53363

@luisherranz Thanks! I rebased on top of that branch and updated the tests. I believe this PR is ready once it's merged assuming all of the tests pass.

SantosGuillamot commented:
> @SantosGuillamot How about we save this for a follow up PR?

That makes sense to me 🙂

artemiomorales committed:
Add clarifying comment to setting of responsive image src attribute

artemiomorales committed:
Update clarifying comment on 1px bug in iOS

SantosGuillamot commented on packages/block-library/src/image/view-interactivity.js:
```diff
@@ -17,293 +17,426 @@ const focusableSelectors = [
+store(
+	{
+		state: {
+			windowWidth: window.innerWidth,
+			windowHeight: window.innerHeight,
```
I don't know if we should use namespaces here as well. It's true that it could be used by any other block, but it could also create conflicts if other blocks declare the same variables.

artemiomorales replied:
@SantosGuillamot Good point — I [pushed a commit](https://github.com/WordPress/gutenberg/pull/52765/commits/b9c84e48b0df2fd0e45588f7eda960132aeed900) adding a namespace.

SantosGuillamot reviewed (approved):
I've tested it in multiple scenarios, and it seems to work great in all of them 👏 I just left a small comment.

artemiomorales committed:
Add namespace to state
